File: src-tauri/src/domain/product.rs
use serde::Serialize;
use time::PrimitiveDateTime;

pub type ProductId = i64;
pub type ProductName = String;
pub type ProductCode = String;
pub type ProductUnit = String;
pub type ProductDefaultPrice = i64;
pub type ProductStandardStockQuantity = i64;

#[derive(Serialize, Debug)]
pub struct Product {
    id: ProductId,
    name: ProductName,
    code: ProductCode,
    unit: ProductUnit,
    default_price: ProductDefaultPrice,
    standard_stock_quantity: ProductStandardStockQuantity,
    created_at: PrimitiveDateTime,
    updated_at: PrimitiveDateTime,
    deleted_at: Option<PrimitiveDateTime>,
}
impl Product {
    pub fn new(
        id: ProductId,
        name: ProductName,
        code: ProductCode,
        unit: ProductUnit,
        default_price: ProductDefaultPrice,
        standard_stock_quantity: ProductStandardStockQuantity,
        created_at: PrimitiveDateTime,
        updated_at: PrimitiveDateTime,
        deleted_at: Option<PrimitiveDateTime>,
    ) -> Self {
        Self {
            id,
            name,
            code,
            unit,
            default_price,
            standard_stock_quantity,
            created_at,
            updated_at,
            deleted_at,
        }
    }

    pub fn id(&self) -> &i64 {
        &self.id
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn code(&self) -> &str {
        &self.code
    }

    pub fn unit(&self) -> &str {
        &self.unit
    }

    pub fn default_price(&self) -> &i64 {
        &self.default_price
    }

    pub fn standard_stock_quantity(&self) -> &i64 {
        &self.standard_stock_quantity
    }

    pub fn created_at(&self) -> &PrimitiveDateTime {
        &self.created_at
    }

    pub fn updated_at(&self) -> &PrimitiveDateTime {
        &self.updated_at
    }

    pub fn deleted_at(&self) -> &Option<PrimitiveDateTime> {
        &self.deleted_at
    }
}

File: src/api/product.ts
import { invoke } from '@/utils/tauri'

export type Product = {
  id: {
    value: number
  }
  name: {
    value: string
  }
  code: {
    value: string
  }
  unit: string
  default_price: number
  standard_stock_quantity: number
  created_at: string
  updated_at: string
  deleted_at?: string
}

type ProductSearchParameters = {
  name?: string
  code?: string
  offset?: number
  limit?: number
}

type FindByIdProductResponse = {
  product: Omit<Product, 'id' | 'name' | 'code'> & {
    id: string
    name: string
    code: string
  }
}

export const findByIdProduct = (productId: number) => {
  return invoke<FindByIdProductResponse>('find_by_id_product', {
    request: {
      product_id: productId,
    },
  })
}

type ProductSearchResponse = {
  products: Product[]
}

export const searchProduct = (params: ProductSearchParameters) => {
  return invoke<ProductSearchResponse>('search_product', {
    request: params,
  })
}

type ProductCreateParameters = {
  name: Product['name']['value']
  code: Product['code']['value']
  unit: Product['unit']
  default_price: Product['default_price']
  standard_stock_quantity: Product['standard_stock_quantity']
}

export const createProduct = (params: ProductCreateParameters) => {
  return invoke('create_product', {
    request: params,
  })
}

type ProductUpdateParameters = {
  id: Product['id']['value']
  name?: Product['name']['value']
  code?: Product['code']['value']
  default_price?: Product['default_price']
  standard_stock_quantity?: Product['standard_stock_quantity']
}

export const updateProduct = (params: ProductUpdateParameters) => {
  return invoke('update_product', {
    request: params,
  })
}
